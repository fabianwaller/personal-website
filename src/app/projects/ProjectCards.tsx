"use client";
import { Button } from "@/components/ui/button";
import {
  Card,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "@/components/ui/card";
import VStack from "@/components/VStack";
import HStack from "@/components/HStack";
import Link from "next/link";
import { Calendar, ExternalLink, GitFork, Globe, Star } from "lucide-react";

import { FaGithub as Github } from "react-icons/fa6";
import { Badge } from "@/components/ui/badge";
import ScrollAnimated from "@/components/ScrollAnimated";
import { Glow, GlowArea } from "@/components/glow";

const formatDate = (date: Date) => {
  return (
    date.getDate() + "." + (date.getMonth() + 1) + "." + date.getFullYear()
  );
};

const ProjectCards = ({ data }: { data: any[] }) => {
  return (
    <GlowArea className={"flex grid-cols-2 flex-col gap-6 md:grid"}>
      {data.map((repo) => {
        if (repo.full_name == "fabianwaller/fabianwaller") return;
        const topics: any[] = repo.topics;
        return (
          <Glow
            key={repo.full_name}
            color="hsl(var(--accent))"
            className="h-full rounded-xl"
          >
            <ScrollAnimated className="h-full">
              <Card className="h-full">
                <CardHeader>
                  <VStack className="items-start">
                    <div className="pt-2 text-sm text-text-light">
                      {repo.type != "closed" ? (
                        <HStack className="mb-0 gap-2 text-sm">
                          <Github className="h-4 w-4" />
                          <span>GitHub</span>
                        </HStack>
                      ) : (
                        <HStack className="mb-0 gap-2 text-sm">
                          <Globe className="h-4 w-4" />
                          <span>Web</span>
                        </HStack>
                      )}
                    </div>
                    <VStack className="items-start" narrow>
                      <CardTitle>{repo.full_name}</CardTitle>
                      <CardDescription>{repo.description}</CardDescription>
                    </VStack>
                    {topics.length > 0 && (
                      <HStack className="gap-2">
                        {topics.map((topic) => (
                          <Badge key={topic} variant={"uncolored"}>
                            {topic}
                          </Badge>
                        ))}
                      </HStack>
                    )}
                  </VStack>
                </CardHeader>
                <CardFooter>
                  <HStack className="mb-0 h-full w-full items-end">
                    <HStack className="w-full items-center justify-between">
                      <HStack className="gap-4 text-sm text-text-light">
                        {repo.type != "closed" && (
                          <HStack className="gap-1">
                            <Star className="h-4 w-4" />
                            {repo.stargazers_count}
                          </HStack>
                        )}
                        {repo.pushed_at && (
                          <HStack className="gap-1">
                            <Calendar className="h-4 w-4" />
                            {formatDate(new Date(repo.pushed_at))}
                          </HStack>
                        )}
                        {repo.fork && (
                          <HStack className="gap-1">
                            <GitFork className="h-4 w-4" />
                          </HStack>
                        )}
                      </HStack>
                      <Link href={repo.html_url} target="_blank">
                        <Button variant={"ghost"} className="left-2 ml-auto">
                          <ExternalLink className="h-4 w-4" />
                        </Button>
                      </Link>
                    </HStack>
                  </HStack>
                </CardFooter>
              </Card>
            </ScrollAnimated>
          </Glow>
        );
      })}
    </GlowArea>
  );
};

export default ProjectCards;
