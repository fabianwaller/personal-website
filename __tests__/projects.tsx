import ProjectCards from "@/app/(website)/projects/ProjectCards";
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";

export const mockProjectData = [
  {
    id: 260796404,
    node_id: "MDEwOlJlcG9zaXRvcnkyNjA3OTY0MDQ=",
    name: "Aura",
    full_name: "fabianwaller/Aura",
    private: false,
    owner: {
      login: "fabianwaller",
      id: 62184437,
      node_id: "MDQ6VXNlcjYyMTg0NDM3",
      avatar_url: "https://avatars.githubusercontent.com/u/62184437?v=4",
      gravatar_id: "",
      url: "https://api.github.com/users/fabianwaller",
      html_url: "https://github.com/fabianwaller",
      followers_url: "https://api.github.com/users/fabianwaller/followers",
      following_url:
        "https://api.github.com/users/fabianwaller/following{/other_user}",
      gists_url: "https://api.github.com/users/fabianwaller/gists{/gist_id}",
      starred_url:
        "https://api.github.com/users/fabianwaller/starred{/owner}{/repo}",
      subscriptions_url:
        "https://api.github.com/users/fabianwaller/subscriptions",
      organizations_url: "https://api.github.com/users/fabianwaller/orgs",
      repos_url: "https://api.github.com/users/fabianwaller/repos",
      events_url: "https://api.github.com/users/fabianwaller/events{/privacy}",
      received_events_url:
        "https://api.github.com/users/fabianwaller/received_events",
      type: "User",
      site_admin: false,
    },
    html_url: "https://github.com/fabianwaller/Aura",
    description: "AuraPvP plugin",
    fork: false,
    url: "https://api.github.com/repos/fabianwaller/Aura",
    forks_url: "https://api.github.com/repos/fabianwaller/Aura/forks",
    keys_url: "https://api.github.com/repos/fabianwaller/Aura/keys{/key_id}",
    collaborators_url:
      "https://api.github.com/repos/fabianwaller/Aura/collaborators{/collaborator}",
    teams_url: "https://api.github.com/repos/fabianwaller/Aura/teams",
    hooks_url: "https://api.github.com/repos/fabianwaller/Aura/hooks",
    issue_events_url:
      "https://api.github.com/repos/fabianwaller/Aura/issues/events{/number}",
    events_url: "https://api.github.com/repos/fabianwaller/Aura/events",
    assignees_url:
      "https://api.github.com/repos/fabianwaller/Aura/assignees{/user}",
    branches_url:
      "https://api.github.com/repos/fabianwaller/Aura/branches{/branch}",
    tags_url: "https://api.github.com/repos/fabianwaller/Aura/tags",
    blobs_url: "https://api.github.com/repos/fabianwaller/Aura/git/blobs{/sha}",
    git_tags_url:
      "https://api.github.com/repos/fabianwaller/Aura/git/tags{/sha}",
    git_refs_url:
      "https://api.github.com/repos/fabianwaller/Aura/git/refs{/sha}",
    trees_url: "https://api.github.com/repos/fabianwaller/Aura/git/trees{/sha}",
    statuses_url:
      "https://api.github.com/repos/fabianwaller/Aura/statuses/{sha}",
    languages_url: "https://api.github.com/repos/fabianwaller/Aura/languages",
    stargazers_url: "https://api.github.com/repos/fabianwaller/Aura/stargazers",
    contributors_url:
      "https://api.github.com/repos/fabianwaller/Aura/contributors",
    subscribers_url:
      "https://api.github.com/repos/fabianwaller/Aura/subscribers",
    subscription_url:
      "https://api.github.com/repos/fabianwaller/Aura/subscription",
    commits_url: "https://api.github.com/repos/fabianwaller/Aura/commits{/sha}",
    git_commits_url:
      "https://api.github.com/repos/fabianwaller/Aura/git/commits{/sha}",
    comments_url:
      "https://api.github.com/repos/fabianwaller/Aura/comments{/number}",
    issue_comment_url:
      "https://api.github.com/repos/fabianwaller/Aura/issues/comments{/number}",
    contents_url:
      "https://api.github.com/repos/fabianwaller/Aura/contents/{+path}",
    compare_url:
      "https://api.github.com/repos/fabianwaller/Aura/compare/{base}...{head}",
    merges_url: "https://api.github.com/repos/fabianwaller/Aura/merges",
    archive_url:
      "https://api.github.com/repos/fabianwaller/Aura/{archive_format}{/ref}",
    downloads_url: "https://api.github.com/repos/fabianwaller/Aura/downloads",
    issues_url:
      "https://api.github.com/repos/fabianwaller/Aura/issues{/number}",
    pulls_url: "https://api.github.com/repos/fabianwaller/Aura/pulls{/number}",
    milestones_url:
      "https://api.github.com/repos/fabianwaller/Aura/milestones{/number}",
    notifications_url:
      "https://api.github.com/repos/fabianwaller/Aura/notifications{?since,all,participating}",
    labels_url: "https://api.github.com/repos/fabianwaller/Aura/labels{/name}",
    releases_url:
      "https://api.github.com/repos/fabianwaller/Aura/releases{/id}",
    deployments_url:
      "https://api.github.com/repos/fabianwaller/Aura/deployments",
    created_at: "2020-05-02T23:52:53Z",
    updated_at: "2024-01-17T11:02:22Z",
    pushed_at: "2020-05-08T17:48:47Z",
    git_url: "git://github.com/fabianwaller/Aura.git",
    ssh_url: "git@github.com:fabianwaller/Aura.git",
    clone_url: "https://github.com/fabianwaller/Aura.git",
    svn_url: "https://github.com/fabianwaller/Aura",
    homepage: null,
    size: 27,
    stargazers_count: 0,
    watchers_count: 0,
    language: "Java",
    has_issues: true,
    has_projects: true,
    has_downloads: true,
    has_wiki: true,
    has_pages: false,
    has_discussions: false,
    forks_count: 0,
    mirror_url: null,
    archived: true,
    disabled: false,
    open_issues_count: 1,
    license: null,
    allow_forking: true,
    is_template: false,
    web_commit_signoff_required: false,
    topics: [],
    visibility: "public",
    forks: 0,
    open_issues: 1,
    watchers: 0,
    default_branch: "master",
  },
];

describe("Projects", () => {
  it("renders projects page unchanged", async () => {
    const component = render(<ProjectCards data={mockProjectData} />);
    expect(component.asFragment()).toMatchSnapshot();
  });
});
