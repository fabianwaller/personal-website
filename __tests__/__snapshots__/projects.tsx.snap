// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects renders projects page unchanged 1`] = `
<DocumentFragment>
  <div
    class="flex grid-cols-2 flex-col gap-6 md:grid"
    style="position: relative; --glow-size: 300px;"
  >
    <div
      class="h-full rounded-lg relative"
      style="--glow-top: 0px; --glow-left: 0px;"
    >
      <div
        class="h-full rounded-lg after:inset-0.25 pointer-events-none absolute inset-0 mix-blend-soft-light after:absolute after:rounded-[inherit] after:bg-background/90 after:content-['']"
      />
      <div
        class="h-full"
        style="opacity: 0; transform: translateY(10px);"
      >
        <div
          class="shadow-custom-soft grid gap-4 overflow-hidden rounded-lg bg-card px-6 py-4 h-full"
        >
          <div
            class="flex flex-col"
          >
            <div
              class="flex flex-col justify-center gap-4 items-start"
            >
              <div
                class="pt-2 text-sm text-text-light"
              >
                <div
                  class="flex flex-row flex-wrap items-center justify-start mb-0 gap-2 text-sm"
                >
                  <svg
                    class="h-4 w-4"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 496 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"
                    />
                  </svg>
                  <span>
                    GitHub
                  </span>
                </div>
              </div>
              <div
                class="flex flex-col justify-center gap-2 items-start"
              >
                <h3
                  class="text-lg font-medium leading-none tracking-tight text-title-normal"
                >
                  fabianwaller/Aura
                </h3>
                <p
                  class="text-normal"
                >
                  AuraPvP plugin
                </p>
              </div>
            </div>
          </div>
          <div
            class="flex items-center"
          >
            <div
              class="flex flex-row flex-wrap justify-start gap-x-4 gap-y-3 mb-0 h-full w-full items-end"
            >
              <div
                class="flex flex-row flex-wrap gap-x-4 gap-y-3 w-full items-center justify-between"
              >
                <div
                  class="flex flex-row flex-wrap items-center justify-start gap-4 text-sm text-text-light"
                >
                  <div
                    class="flex flex-row flex-wrap items-center justify-start gap-1"
                  >
                    <svg
                      class="lucide lucide-star h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
                      />
                    </svg>
                    0
                  </div>
                  <div
                    class="flex flex-row flex-wrap items-center justify-start gap-1"
                  >
                    <svg
                      class="lucide lucide-calendar h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M8 2v4"
                      />
                      <path
                        d="M16 2v4"
                      />
                      <rect
                        height="18"
                        rx="2"
                        width="18"
                        x="3"
                        y="4"
                      />
                      <path
                        d="M3 10h18"
                      />
                    </svg>
                    8.5.2020
                  </div>
                </div>
                <a
                  href="https://github.com/fabianwaller/Aura"
                  target="_blank"
                >
                  <button
                    class="relative inline-flex items-center justify-center gap-1 whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 hover:bg-card-foreground hover:text-primary-foreground left-2 ml-auto"
                  >
                    <svg
                      class="lucide lucide-external-link h-4 w-4"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M15 3h6v6"
                      />
                      <path
                        d="M10 14 21 3"
                      />
                      <path
                        d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
                      />
                    </svg>
                  </button>
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
